appinventor/components/src/com/google/appinventor/components/runtime/Ev3UltrasonicSensor.java

```diff
@@ -255,65 +252,63 @@ public void AboveRange() {
    */
   @DesignerProperty(editorType = PropertyTypeConstants.PROPERTY_TYPE_LEGO_EV3_ULTRASONIC_SENSOR_MODE,
-                    defaultValue = DEFAULT_SENSOR_MODE_STRING)
+      defaultValue = "cm")
   @SimpleProperty
-  public void Unit(String unitName) {
-    String functionName = "Unit";
-    try {
-      setMode(unitName);
-    } catch(IllegalArgumentException e) {
-      form.dispatchErrorOccurredEvent(this, functionName, ErrorMessages.ERROR_EV3_ILLEGAL_ARGUMENT, functionName);
+  public void Unit(@Options(UltrasonicSensorUnit.class) String unitName) {
+    // Make sure unitName is a valid UltrasonicSensorUnit.
+    UltrasonicSensorUnit unit = UltrasonicSensorUnit.fromUnderlyingValue(unitName);
+    if (unit == null) {
+      form.dispatchErrorOccurredEvent(
+        this, "Unit", ErrorMessages.ERROR_EV3_ILLEGAL_ARGUMENT, unitName);
+      return;
     }
+    setMode(unit);
+  }
+
+  /**
+   * Sets the unit of this ultrasonic sensor.
+   */
+  public void UnitAbstract(UltrasonicSensorUnit unit) {
+    setMode(unit);
   }
 
   /**
    * Returns the unit of distance.
    */
   @SimpleProperty(description = "The distance unit, which can be either \"cm\" or \"inch\".",
                   category = PropertyCategory.BEHAVIOR)
-  public String Unit() {
-    return modeString;
+  public @Options(UltrasonicSensorUnit.class) String Unit() {
+    return mode.toUnderlyingValue();
+  }
+
+  /**
+   * Returns the current unit of distance for this ultrasonic sensor.
+   */
+  public UltrasonicSensorUnit UnitAbstract() {
+    return mode;
   }
 
   /**
    * Measure the distance in centimeters.
    */
   @SimpleFunction(description = "Measure the distance in centimeters.")
+  @Deprecated
   public void SetCmUnit() {
-    String functionName = "SetCmUnit";
-    try {
-      setMode(SENSOR_MODE_CM_STRING);
-    } catch(IllegalArgumentException e) {
-      form.dispatchErrorOccurredEvent(this, functionName, ErrorMessages.ERROR_EV3_ILLEGAL_ARGUMENT, functionName);
-    }
+    setMode(UltrasonicSensorUnit.Centimeters);
   }
 
   /**
    * Measure the distance in inches.
    */
   @SimpleFunction(description = "Measure the distance in inches.")
+  @Deprecated
   public void SetInchUnit() {
-    String functionName = "SetInchUnit";
-    try {
-      setMode(SENSOR_MODE_INCH_STRING);
-    } catch(IllegalArgumentException e) {
-      form.dispatchErrorOccurredEvent(this, functionName, ErrorMessages.ERROR_EV3_ILLEGAL_ARGUMENT, functionName);
-    }
+    setMode(UltrasonicSensorUnit.Inches);
   }
 
-  private void setMode(String newModeString) {
+  private void setMode(UltrasonicSensorUnit mode) {
     previousDistance = -1.0;
-
-    if (SENSOR_MODE_CM_STRING.equals(newModeString)) {
-      mode = SENSOR_MODE_CM;
-    }
-    else if (SENSOR_MODE_INCH_STRING.equals(newModeString)) {
-      mode = SENSOR_MODE_INCH;
-    }
-    else
-      throw new IllegalArgumentException();
-
-    this.modeString = newModeString;
+    mode = mode;
```

Comment:
Need to use `this.mode` on the left hand side, otherwise this is a no-op assigning the parameter mode to itself. Alternatively, rename the parameter to newMode similar to the other files.